ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveInBetweenExpandRule.java

```diff
@@ -0,0 +1,198 @@
+
+      if (!expander.modified) {
+        return;
+      }
+
+      RelNode newJoin = join.copy(join.getTraitSet(),
+          condition,
+          join.getLeft(),
+          join.getRight(),
+          join.getJoinType(),
+          join.isSemiJoinDone());
+
+      call.transformTo(newJoin);
+    }
+  }
+
+  /** Rule adapter to apply the transformation to Project expressions. */
+  private static class ProjectRule extends RelOptRule {
+
+    ProjectRule() {
+      super(operand(Project.class, any()), HiveRelFactories.HIVE_BUILDER, null);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+      final Project project = call.rel(0);
+      RexInBetweenExpander expander = new RexInBetweenExpander(
+          project.getCluster().getRexBuilder());
+      List<RexNode> newProjects = new ArrayList<>();
+      for (RexNode expr : project.getProjects()) {
+        newProjects.add(expander.apply(expr));
+      }
+
+      if (!expander.modified) {
+        return;
+      }
+
+      Project newProject = project.copy(project.getTraitSet(),
+          project.getInput(), newProjects, project.getRowType());
+
+      call.transformTo(newProject);
+    }
+  }
+
+
+  /**
+   * Class that transforms IN/BETWEEN clauses in an expression.
+   * If any call is modified, the modified flag will be set to
+   * true after its execution.
+   */
+  private static final class RexInBetweenExpander extends RexShuttle {
+
+    private final RexBuilder rexBuilder;
+    private boolean modified;
+
+    private RexInBetweenExpander(RexBuilder rexBuilder) {
+      this.rexBuilder = rexBuilder;
+      this.modified = false;
+    }
+
+    @Override
+    public RexNode visitCall(final RexCall call) {
+      switch (call.getKind()) {
+      case AND: {
+        boolean[] update = {false};
+        List<RexNode> newOperands = visitList(call.operands, update);
+        if (update[0]) {
+          return RexUtil.composeConjunction(rexBuilder, newOperands);
+        }
+        return call;
+      }
+      case OR: {
+        boolean[] update = {false};
+        List<RexNode> newOperands = visitList(call.operands, update);
+        if (update[0]) {
+          return RexUtil.composeDisjunction(rexBuilder, newOperands);
+        }
+        return call;
+      }
+      case IN: {
+        List<RexNode> newOperands = RexNodeConverter.transformInToOrOperands(
+            call.getOperands(), rexBuilder);
+        if (newOperands == null) {
+          // We could not execute transformation, return expression
+          return call;
+        }
+        modified = true;
+        if (newOperands.size() > 1) {
+          return rexBuilder.makeCall(SqlStdOperatorTable.OR, newOperands);
+        }
+        return newOperands.get(0);
+      }
+      case BETWEEN: {
+        List<RexNode> newOperands = RexNodeConverter.rewriteBetweenChildren(
+            call.getOperands(), rexBuilder);
+        modified = true;
+        if (call.getOperands().get(0).isAlwaysTrue()) {
```

Comment:
I am not sure if I understand this. In what case BETWEEN is rewritten into OR?

Reply:
The first parameter to BETWEEN is a boolean flag, to differentiate BETWEEN and NOT BETWEEN.
`a BETWEEN b AND c` => `b <= a AND a <= c`
`a NOT BETWEEN b AND c` => `b > a OR a > c`